Built in profiling

BaronKiko opened the pull request:
This PR attempts to implement a profiler within ryujinx.  Currently it only profiles system services and I want to know what you want profiled. It's only a couple lines to add one so don't be shy.
Honestly it's not the nicest UI but it's pretty lightweight and not really intended for end users but devs to be able to find potential bottlenecks.

It looks like this (slightly out of date but you get the idea):
![ryujinx_2019-01-26_21-14-27](https://user-images.githubusercontent.com/16206533/51792872-85d78c00-21af-11e9-8240-3dadfc85ec45.png)
![2019-01-29_04-42-42](https://user-images.githubusercontent.com/16206533/51884769-9ee36700-2380-11e9-9d6b-aa6d81d71966.gif)

Features (so you know what to test):

- F1 (change in config) to hide/show the profiler. Only works when main window focused
- Click column heading to sort
- Scrollable data
- Hide inactive button
- Play/Pause button
- Step button when paused
- Filtering, type to change filter with profiler focused
- Regex filtering, click filter bar at bottom to toggle
- Entries should be searchable as Category.Group.Item. For example you can get only the top entry in the picture above with "ServiceCall.ISharedFontManager.GetFontSize" or emit GetFontSize to get all children of ISharedFontManager
- Timing bars render correctly (you have to switch to it with view button)
- No idea if mac or linux work (can someone who has them set up test?)
- Changing update rate from config
- CSV dumping, only enabled with a valid path specified in config.
- Graphing, you should be able to scroll/zoom with arrow
- Flags, the vertical lines that appear in graph mode. They represent a frame swap atm but you can add more later. Should probably draw them in different colours but there is only 1 at the time of writing this.
- Maybe some other things, hard to remember all I did


Known issues:

- Need a better way to get default font locations

I think that's all I have to say atm. Hopefully it works for you.

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+            int linePadding = 2;
+            int columnSpacing = 30;
+            int filterHeight = 24;
+
```
align, below too

emmauss replied:
align, below too

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+            for (int i = 0; i < ProfileData.Count; i += 2)
+            {
+                float top = GetLineY(yOffset, lineHeight, linePadding, false, i - 1);
+                float bottom = GetLineY(yOffset, lineHeight, linePadding, false, i);
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+                maxWidth = width;
+
+            xOffset += maxWidth + columnSpacing;
+            
```
remove whitespace

emmauss replied:
remove whitespace

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+
+            // Display session group
+            maxWidth = 0;
+            verticalIndex = 0;
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+
+            // Display session item
+            maxWidth = 0;
+            verticalIndex = 0;
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+            if (ProfileData.Count != 0)
+            {
+                width = Width - xOffset - 370;
+                long maxAverage, maxTotal;
```
declare these at the beginning of this scope,  each variable on a new line

emmauss replied:
declare these at the beginning of this scope,  each variable on a new line

emmauss replied:
declare these at the beginning of this scope,  each variable on a new line

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+                {
+                    maxInstant = Math.Max(maxInstant, kvp.Value.Instant);
+                    maxAverage = Math.Max(maxAverage, kvp.Value.AverageTime);
+                    maxTotal = Math.Max(maxTotal, kvp.Value.TotalTime);
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,545 @@
+            }
+        }
+
+        protected override void OnMouseMove(MouseMoveEventArgs e)
```
if not abstract, remove

emmauss replied:
if not abstract, remove

emmauss commented on Ryujinx.Profiler/UI/ProfileWindowManager.cs:
```diff
@@ -0,0 +1,44 @@
+    public class ProfileWindowManager
+    {
+        private ProfileWindow Window;
+        private Thread ProfileThread;
```
_camelCase for private fields

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+            public float left, right, bottom, top;
+            public float height, aspectRatio, xBearing, yBearing, advance;
+            public int width;
+        }
```
each variable on new line

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+        {
+            // Create and init some vars
+            uint[] rawCharacterSheet = new uint[SheetWidth * SheetHeight];
+            int x, y, lineOffset, maxHeight;
```
variable per line

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+            // Update raw data for each character
+            for (int i = 0; i < 94; i++)
+            {
+                float xBearing, yBearing, advance;
```
this

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+                characters[i].yBearing /= maxHeight;
+                characters[i].advance /= characters[i].width;
+                characters[i].height /= maxHeight;
+                characters[i].aspectRatio = (float)characters[i].width / maxHeight;
```
align

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMinFilter, (int)TextureMinFilter.Linear);
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMagFilter, (int)TextureMagFilter.Linear);
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapS, (int)TextureWrapMode.Clamp);
```
align end

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+            var glyph = font.GetGlyph(c, 32);
+            xBearing = glyph.HorizontalMetrics.Bearing.X;
+            yBearing = glyph.RenderHeight - glyph.HorizontalMetrics.Bearing.Y;
+            advance = glyph.HorizontalMetrics.Advance;
```
align

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+                Bits = Marshal.AllocHGlobal(glyph.RenderWidth * glyph.RenderHeight),
+                Width = glyph.RenderWidth,
+                Height = glyph.RenderHeight,
+                Pitch = glyph.RenderWidth
```
align

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+            int width = surface.Width;
+            int height = surface.Height;
+            int len = width * height;
+            byte[] data = new byte[len];
```
align

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+            for (int y = 0; y < height; y++)
+            {
+                int destOffset = (y + posY) * SheetWidth + posX;
+                int sourceOffset = y * width;
```
align

emmauss commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,230 @@
+                bottom = (float)(posY + height) / SheetHeight,
+                width = width,
+                height = height,
+            };
```
align

BaronKiko commented:
@emmauss Waiting for that :) There is enough code that it will probably take a couple passes. Anyway I will review your suggestions after dinner, and maybe sleep.

BaronKiko commented on Ryujinx.Profiler/InternalProfile.cs:
```diff
@@ -0,0 +1,105 @@
+        internal ConcurrentDictionary<ProfileConfig, TimingInfo> Timers;
+
+        private readonly object SessionLock = new object();
+        private int SessionCounter = 0;
```
Internal should be Pascal according to Gdk, otherwise done. Commit coming later,

BaronKiko commented on Ryujinx.Profiler/Profile.cs:
```diff
@@ -0,0 +1,86 @@
+                method();
+
+            Begin(config);
+            method();
```
> pascal case for method. If there is a naming conflict, think of a better name. `method` is too generic

Not exactly sure what you mean here.
The idea of this function is it's called with Profie.Method(some_config, () => do something);
So you are profiling a method, it can be anything so the name is meant to be very generic.
If you are referring to the variable it's a local var so lower case is right?

BaronKiko commented on Ryujinx.Profiler/Settings.cs:
```diff
@@ -0,0 +1,14 @@
+        public bool FileDumpEnabled = false;
+        public string DumpLocation = "";
+        public float UpdateRate = 0.1f;
+    }
```
Not sure what you mean by use auto props as they don't really apply here. Those are just the defaults. I have added a comment to clarify that they are defaults and aligned them

jduncanator replied:
General rule is to make any public member a property, those are currently public fields.

BaronKiko commented:
Stopped using window.Run so it should be easier on your cpu/gpu.
It also now only renders a frame when there is actually new data/ui interaction to display
@emmauss let me know how it goes for you, I hope it's worth it because it was a pain

BaronKiko commented:
Huge update, adds real time bar graph support:
![image](https://user-images.githubusercontent.com/16206533/51884563-b110d580-237f-11e9-94a9-88f966c8ffc6.png)
It also supports TimingFlags (the vertical bars), I have currently set them up for ogl frame swap so you can clearly see inconsistent frame times.
You can change update rate to 60fps in config to get proper real time but it may give you performance issues as it's processing a lot of data.
Anyway enjoy

BaronKiko commented:
I think that's basically it for the actual profiler but I still need to know what to profile and ideally sort out the system font issue as looking at the VirtualFileSystem class I'm just more confused. If I understood it correctly no code change is needed.

BaronKiko commented:
Please review for merge :)

emmauss commented on Ryujinx.HLE/HOS/Services/IpcService.cs:
```diff
@@ -95,7 +96,13 @@ public void CallMethod(ServiceCtx context)
 
+                ProfileConfig profile = Profiles.ServiceCall;
+                profile.SessionGroup = service.GetType().Name;
+                profile.SessionItem = processRequest.Method.Name;
```
align

emmauss commented on Ryujinx.Profiler/Settings.cs:
```diff
@@ -0,0 +1,22 @@
+        public int    MaxLevel        { get; set; } = 0;
+
+        // 19531225 = 5 seconds in ticks
+        public long   History         { get; set; } = 19531225;
```
no need for the spaces.

emmauss commented on Ryujinx.Profiler/TimingInfo.cs:
```diff
@@ -0,0 +1,175 @@
+        // Depth of current timer,
+        // each begin call increments and each end call decrements
+        private int _depth;
+        
```
extra new line

emmauss commented on Ryujinx.Profiler/UI/ProfileButton.cs:
```diff
@@ -0,0 +1,111 @@
+        private Action _clicked;
+        private bool   _visible;
+
+
```
extra new line

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,726 @@
+        // UI variables
+        private ProfileButton[] _buttons;
+
+        private bool _initComplete    = false;
```
no  need for the new lines below. the can be grouped together

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,726 @@
+
+        // Input
+        private bool _backspaceDown       = false;
+        private bool _prevBackspaceDown   = false;
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,726 @@
+        private bool   _profileUpdated           = false;
+        private readonly object _profileDataLock = new object();
+        
+
```
no need for a new line

emmauss commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,726 @@
+                // Totals
+                yHeight = FilterHeight + 2;
+                _fontService.DrawText($"{totalInstant / PerformanceCounter.TicksPerMillisecond:F3} ({totalCount})", xOffset, yHeight, TitleFontHeight);
+                _fontService.DrawText($"{totalAverage / PerformanceCounter.TicksPerMillisecond:F3}", 150 + xOffset, yHeight, TitleFontHeight);
```
align the params

emmauss commented on Ryujinx.Profiler/UI/ProfileWindowBars.cs:
```diff
@@ -0,0 +1,83 @@
+                {
+                    maxInstant = Math.Max(maxInstant, kvp.Value.Instant);
+                    maxAverage = Math.Max(maxAverage, kvp.Value.AverageTime);
+                    maxTotal = Math.Max(maxTotal, kvp.Value.TotalTime);
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindowBars.cs:
```diff
@@ -0,0 +1,83 @@
+                    // Instant
+                    GL.Color3(Color.Blue);
+                    float bottom = GetLineY(yOffset, LineHeight, LinePadding, true, verticalIndex++);
+                    float top = bottom + barHeight;
```
align

emmauss commented on Ryujinx.Profiler/UI/ProfileWindowBars.cs:
```diff
@@ -0,0 +1,83 @@
+                    GL.Color3(Color.Red);
+                    top += barHeight;
+                    bottom += barHeight;
+                    right = (float)entry.Value.TotalTime / maxTotal * width + xOffset;
```
align, and maybe a new line after this line.

BaronKiko commented:
Fixed formatting issues left
Added frame timing display to cover this commit closing: https://github.com/Ryujinx/Ryujinx/pull/551
Coloured flags to differentiate between flag types
Show/Hide flag button added
Needs another review

BaronKiko commented:
Bump maybe?

gdkchan commented on Ryujinx.HLE/PerformanceStatistics.cs:
```diff
@@ -1,4 +1,5 @@
-﻿using System.Diagnostics;
+﻿using Ryujinx.Profiler;
+using System.Diagnostics;
```
Is this really needed?

BaronKiko replied:
It's needed for the stopwatch class

gdkchan commented on Ryujinx.Profiler/DumpProfile.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            String fileData = "Category,Session Group,Session Item,Count,Average(ms),Total(ms)\r\n";
+
+            foreach (var time in profile.Timers.OrderBy(key => key.Key.Tag))
```
I don't like the user of var here as its not easy to see the type of `time` without looking at the definition on `InternalProfile`.

gdkchan commented on Ryujinx.Profiler/InternalProfile.cs:
```diff
@@ -0,0 +1,224 @@
+        public string GetSession()
+        {
+            // Can be called from multiple threads so locked to ensure no duplicate sessions are generated
+            lock (_sessionLock)
```
You don't need a lock for that, you can use `Interlocked.Increment` to increment the value atomically.

BaronKiko replied:
Ah cool, I didn't know about that.

gdkchan commented on Ryujinx.Profiler/NPadDebug.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Ryujinx.Profiler
+{
+    public struct NpadDebugButtons
```
Why is it called `Npad`? I know that on the HID service, Npad is usually used to refer to the different controller types, but this seems completely unrelated, so it's not clear to me why it is called `NpadDebug`.

BaronKiko replied:
Fair point. Refactored out

gdkchan commented on Ryujinx.Profiler/Profile.cs:
```diff
@@ -0,0 +1,143 @@
+
+        private static InternalProfile  _profileInstance;
+        private static ProfilerSettings _settings;
+
```
nit: Unneeded empty new line.

gdkchan commented on Ryujinx.Profiler/Profile.cs:
```diff
@@ -0,0 +1,143 @@
+
+        public static bool ProfilingEnabled()
+        {
+            #if USE_PROFILING
```
Usually those #if #else #endif etc directives are not idented. As it is now, it's imo a bit hard to see at a first glance what is supposed to run on each condition.

BaronKiko replied:
I can unindent it. I usually keep them indented just because I think it looks nicer

gdkchan commented on Ryujinx.Profiler/ProfileConfig.cs:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        /// <summary>
+        /// This equals overload is vital
```
I don't think that `This equals overload is vital` is necessary on the summary.

gdkchan commented on Ryujinx.Profiler/ProfileConfig.cs:
```diff
@@ -0,0 +1,121 @@
+        /// </summary>
+        /// <param name="obj">Object to compare to</param>
+        /// <returns></returns>
+        public override bool Equals(object obj)
```
May be worth to implement `IEquatable<ProfileConfig>` and add a equals overload accepting `ProfileConfig` to avoid boxing in some cases.

gdkchan commented on Ryujinx.Profiler/ProfilerConfig.jsonc:
```diff
@@ -0,0 +1,28 @@
+﻿{
+  // Enable profiling (Only available on a profiling enabled build)
```
nit: `profiling enabled build` -> `profiling enabled builds`.

gdkchan commented on Ryujinx.Profiler/Settings.cs:
```diff
@@ -0,0 +1,23 @@
+        public int    MaxFlags        { get; set; } = 1000;
+
+        // 19531225 = 5 seconds in ticks
+        public long   History { get; set; } = 19531225;
```
I beleive the extra spacing between long and History is not necessary. Also I believe the counter frequency *may* vary from system to system. So while "19531225" is equivalent to 5 seconds on your system, it may not be in other systems.

BaronKiko replied:
I fixed the spacing
The time equals that number on any pc with a high precision clock which I have been led to believe is most computers.
It shouldn't really matter anyway because it sets that value on boot with the time specified in config. I have updated the comment to clarify this.

gdkchan commented on Ryujinx.Profiler/TimingFlag.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Ryujinx.Profiler
+{
+    public enum TimingFlagType
```
Why it this enum explicitly numbered?

gdkchan commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,772 @@
+
+            // Update this when new buttons are added
+            Count             = 10,
+            Autodraw          = 8,
```
Was this a mistake? Value 8 is already used by `ToggleFlags`. If not, I think would be nice to add a comment here.

BaronKiko replied:
These 2 values are used as indexes in the list. This is also why they are explicitly numbered.
Up to auto draw it will automatically do the rendering of the buttons with a single loop. After that it draws at various points in the code. This is so it can control visibility of the buttons and update positions.

gdkchan commented on Ryujinx.Profiler/UI/ProfileWindow.cs:
```diff
@@ -0,0 +1,772 @@
+            _profileUpdated = true;
+        }
+
+        #region OnLoad
```
iirc #region is usually also not idented.

gdkchan commented on Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs:
```diff
@@ -0,0 +1,257 @@
+        {
+            string fontFolder = System.Environment.GetFolderPath(Environment.SpecialFolder.Fonts);
+
+            // Only uses Arial, add more font here if wanted
```
`more font` -> `more fonts`.

gdkchan commented on Ryujinx/Program.cs:
```diff
@@ -5,6 +5,7 @@
 using Ryujinx.HLE;
 using System;
 using System.IO;
+using Ryujinx.Profiler;
```
nit: Sorting.

BaronKiko commented:
Fixed all review comments

gdkchan commented on Ryujinx.Profiler/InternalProfile.cs:
```diff
@@ -0,0 +1,221 @@
+        private readonly object _timerQueueClearLock = new object();
+        private ConcurrentQueue<TimerQueueValue> _timerQueue;
+
+        private readonly object _sessionLock = new object();
```
I guess this is now unused?

BaronKiko replied:
Quite right. Removed

BaronKiko commented:
I have tested it on my system. Would be nice to see performance on a weaker system but I really don't see how I could optimize it further so I'm not sure what I could do to help that.
It's all stripped out on normal builds so it shouldn't effect normal users.

I missed these comments
> I see that you wrote a whole GUI using OpenTK for the profiler, which is nice but we're going to need rewriting this portion of the code once we have a GUI on the emulator

It will be tough to keep up for most ui's but we can come to that when a ui happens

>  it is using legacy OpenGL

I probably should have used modern gl but at the time I chose to go for maximum compatibility. Modern ogl wasn't needed and traditional opengl allows some optimizations that would be harder to make, perhaps impossible.
